File: mq/src/main/java/io/dourl/mqtt/ui/ChatActivity.java
package io.dourl.mqtt.ui;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.Arrays;

import io.dourl.mqtt.R;
import io.dourl.mqtt.base.log.LoggerUtil;
import io.dourl.mqtt.bean.MessageModel;
import io.dourl.mqtt.bean.MessagePaging;
import io.dourl.mqtt.bean.SessionModel;
import io.dourl.mqtt.bean.UserModel;
import io.dourl.mqtt.event.ChatMsgEvent;
import io.dourl.mqtt.event.MsgStatusUpdateEvent;
import io.dourl.mqtt.manager.EventBusManager;
import io.dourl.mqtt.manager.MessageManager;
import io.dourl.mqtt.model.message.chat.BodyType;
import io.dourl.mqtt.model.message.chat.TextBody;
import io.dourl.mqtt.model.message.emoji.DefaultEmojiconDatas;
import io.dourl.mqtt.model.message.emoji.Emojicon;
import io.dourl.mqtt.model.message.emoji.EmojiconGroupEntity;
import io.dourl.mqtt.storage.DbCallback;
import io.dourl.mqtt.storage.MessageDao;
import io.dourl.mqtt.storage.SessionDao;
import io.dourl.mqtt.storage.SessionManager;
import io.dourl.mqtt.ui.adpater.chat.ChatAdapter;
import io.dourl.mqtt.ui.adpater.chat.ChatTextProvider;
import io.dourl.mqtt.ui.widge.EmojiconMenu;
import io.dourl.mqtt.ui.widge.EmojiconMenuBase;
import io.dourl.mqtt.ui.widge.MultiLineEditText;
import io.dourl.mqtt.ui.widge.RecordStateView;
import io.dourl.mqtt.ui.widge.RecordView;
import io.dourl.mqtt.ui.widge.TitleBar;
import io.dourl.mqtt.utils.AppContextUtil;
import io.dourl.mqtt.utils.ImSmileUtils;
import io.reactivex.functions.Consumer;


/**
 * File description.
 *
 * @author dourl
 * @date 2022/3/8
 */
public class ChatActivity extends BaseActivity implements View.OnClickListener,RecordView.AudioRecordListener {
    private static final String TAG = "ChatActivity";
    RecyclerView mRecyclerView;
    TextView noticeContent;
    ImageView noticeClose;
    RelativeLayout noticeLayout;
    TitleBar mTitleBar;
    TextView msgCountHint;
    TextView msgUnread;
    MultiLineEditText mEditText;
    ImageView mBtnSend;
    ImageButton btnAlbum;
    ImageButton btnPhoto;
    ImageButton btnRecord;
    ImageButton mBtnFace;
    FrameLayout mEmojiMenuContainer;
    LinearLayout inputView;
    ImageView btnKeyboard;
    RecordView recordView;

    FrameLayout bottomView;
    RecordStateView recordStateView;
    RelativeLayout mMainLayout;

    protected String mSessionID = null;
    private UserModel mBaseUser;
    private ChatAdapter mAdapter;
    protected int mUnreadCount = 15;
    private long mLastMsgId;
    protected TextView mMsgUnread;
    protected boolean ctrlPress = false;
    private boolean rightShown = false;

    public static void intentTo(Context context, String sessionId) {
        Intent intent = new Intent(context, ChatActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("session_id", sessionId);
        intent.putExtras(bundle);
        context.startActivity(intent);
    }

    public static void intentTo(Context context, String sessionId, UserModel user) {
        Intent intent = new Intent(context, ChatActivity.class);
        Bundle bundle = new Bundle();
        bundle.putString("session_id", sessionId);
        bundle.putParcelable("user", user);
        intent.putExtras(bundle);
        context.startActivity(intent);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);
        EventBusManager.getInstance().register(this);
        init();
        initView();

        mTitleBar.showLine();
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        linearLayoutManager.setAutoMeasureEnabled(true);
        linearLayoutManager.setStackFromEnd(true);
        linearLayoutManager.setReverseLayout(true);
        mRecyclerView.setLayoutManager(linearLayoutManager);
        mRecyclerView.setHasFixedSize(false);
        mAdapter = new ChatAdapter(new ArrayList<MessageModel>());
        mAdapter.register(TextBody.class, new ChatTextProvider());
        mRecyclerView.setAdapter(mAdapter);
        setupMessageList();
        setupInputBar();
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (linearLayoutManager.findLastCompletelyVisibleItemPosition() == linearLayoutManager.getItemCount() - 1) {
                    LoggerUtil.d(TAG,"reach top");
//                    mPtrLayout.autoRefresh(true, 200);
                    getMsgsByPage(0);
                }
                if (mMsgUnread.isShown() && linearLayoutManager.findLastCompletelyVisibleItemPosition() == mUnreadCount - 1) {
                    mMsgUnread.setVisibility(View.GONE);
                }
            }
        });
        mRecyclerView.addOnLayoutChangeListener(new View.OnLayoutChangeListener() {
            @Override
            public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                if (bottom < oldBottom - 300) {
                    LoggerUtil.d(TAG,"LayoutChange oldBottom");
                    mRecyclerView.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            scrollToBottom();
                        }
                    }, 100);
                }
            }
        });
        initEmoji();

    }
    protected EmojiconMenuBase emojiconMenu;

    private void initEmoji() {
        if (emojiconMenu == null) {
            emojiconMenu = (EmojiconMenu) LayoutInflater.from(this).inflate(R.layout.im_layout_emojicon_menu, null);
            ArrayList<EmojiconGroupEntity> emojiconGroupList = new ArrayList<EmojiconGroupEntity>();
            emojiconGroupList.add(new EmojiconGroupEntity(R.drawable.ee_1, Arrays.asList(DefaultEmojiconDatas.getData())));
            ((EmojiconMenu) emojiconMenu).init(emojiconGroupList);
        }
        mEmojiMenuContainer.addView(emojiconMenu);

        // emojicon menu
        emojiconMenu.setEmojiconMenuListener(new EmojiconMenuBase.EmojiconMenuListener() {

            @Override
            public void onExpressionClicked(Emojicon emojicon) {
                if (emojicon.getType() != Emojicon.Type.BIG_EXPRESSION) {
                    if (emojicon.getEmojiText() != null) {
                        mEditText.append(ImSmileUtils.getSmiledText(ChatActivity.this, emojicon.getEmojiText()));
                    }
                } else {
//                    sendBigExpressionMessage(emojicon.getName(), emojicon.getIdentityCode());
                }
            }

            @Override
            public void onDeleteImageClicked() {
                if (!TextUtils.isEmpty(mEditText.getText())) {
                    KeyEvent event = new KeyEvent(0, 0, 0, KeyEvent.KEYCODE_DEL, 0, 0, 0, 0, KeyEvent.KEYCODE_ENDCALL);
                    dispatchKeyEvent(event);
                }
            }

            @Override
            public void onSendClicked() {
                sendTxtMessage();
            }
        });
    }

    protected void toggleEmojicon() {
        if (mEmojiMenuContainer.getVisibility() == View.VISIBLE) {
            mEmojiMenuContainer.setVisibility(View.GONE);
        } else {
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    mEmojiMenuContainer.setVisibility(View.VISIBLE);
                }
            }, 50);
            hideKeyboard();
        }

    }
    private void setupInputBar() {
        mEditText.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {

                // test on Mac virtual machine: ctrl map to KEYCODE_UNKNOWN
                if (keyCode == KeyEvent.KEYCODE_UNKNOWN) {
                    if (event.getAction() == KeyEvent.ACTION_DOWN) {
                        ctrlPress = true;
                    } else if (event.getAction() == KeyEvent.ACTION_UP) {
                        ctrlPress = false;
                    }
                }
                return false;
            }
        });

        mEditText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_SEND || actionId == EditorInfo.IME_ACTION_DONE ||
                        (event.getKeyCode() == KeyEvent.KEYCODE_ENTER &&
                                event.getAction() == KeyEvent.ACTION_DOWN &&
                                ctrlPress == true)) {
                    sendTxtMessage();
                    return true;
                }

                return false;
            }
        });
        mEditText.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (b) {
                    toggleEditMode();
                }
            }
        });

        mRecyclerView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_UP) {
                    hideKeyboard();
                    if (rightShown)
                        toggleRightLayout();
                }
                return false;
            }
        });
    }

    protected void toggleRightLayout() {
        ObjectAnimator animator = null;
        if (rightShown) {
            animator = ObjectAnimator.ofFloat(mMainLayout, "translationX", -AppContextUtil.dip2px(240), 0f);
        } else {
            animator = ObjectAnimator.ofFloat(mMainLayout, "translationX", 0f, -AppContextUtil.dip2px(240));
        }
        animator.setDuration(500);
        animator.setInterpolator(new AccelerateDecelerateInterpolator());
        animator.start();
        rightShown = !rightShown;
        hideKeyboard();
    }


    protected void hideKeyboard() {
        final View focusView = getCurrentFocus();
        if (focusView != null) {
            InputMethodManager imm = (InputMethodManager) focusView.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(focusView.getWindowToken(), 0);
            focusView.clearFocus();
        }
    }

    private void toggleEditMode() {

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                mEmojiMenuContainer.setVisibility(View.GONE);
            }
        }, 50);
    }

    private void sendTxtMessage() {
        String text = mEditText.getText().toString();
        if (text != null && text.length() > 0) {
            MessageManager.getInstance().sendTextMessage(mBaseUser, text);
            mEditText.setText("");
            scrollToBottom();
        } else {
            // sendEmptyTxtMessage();
        }
        if (!mBaseUser.isFriend()) {
            /*HINT_LIMIT--;
            if (HINT_LIMIT == -1) {
                new ChatHintDialog(this, mMainLayout, new ChatHintDialog.OnFunClickListener() {
                    @Override
                    public void onFunClick() {
                        showLoadingProgress();
                        doContactsAction();
                    }
                }).show();
            }*/
        }
    }

    private void scrollToBottom() {
        mRecyclerView.scrollToPosition(0);
    }

    private void setupMessageList() {
        SessionDao.findSessionById(mSessionID, new DbCallback<SessionModel>() {
            @Override
            public void onSuccess(SessionModel model) {

                runOnUiThread(() -> {
                    if (model != null) {
                        mUnreadCount = model.getUnreadMsgCount();
                        if (mUnreadCount > 15) {
                            mMsgUnread.setVisibility(View.VISIBLE);
                            mMsgUnread.setText(getString(R.string.hint_chat_unread_msg, mUnreadCount));
                        }
                          mEditText.setText(model.getDraft());
                    }
                    getMsgsByPage(mUnreadCount);
                    SessionManager.getInstance().resetUnreadCount(mSessionID);
                });


            }

            @Override
            public void onFail(Throwable e) {

            }
        });

    }
    public static final int PRE_PAGE_COUNT = 10; //最少10
    @SuppressLint("CheckResult")
    private void getMsgsByPage(int unread) {
        MessageDao.getMsgsBySessionIdObservable(mSessionID, mLastMsgId, PRE_PAGE_COUNT + unread).subscribe(new Consumer<MessagePaging>() {
            @Override
            public void accept(@NonNull MessagePaging messagePaging) throws Exception {
                if (messagePaging == null) return;
                if (0 == mLastMsgId) {
                    mAdapter.setData(messagePaging.list);
                    scrollToBottom();
//                    mRecyclerView.scrollToPosition(messagePaging.list.size() - 1);
                } else {
                    mAdapter.addPreviousPage(messagePaging.list);
                }
                mLastMsgId = messagePaging.lastMsgDbId;
                //mPtrLayout.setPullToRefresh(messagePaging.hasMore);
            }

        });
    }

    protected void initView() {
        mTitleBar = findViewById(R.id.titleBar);
        mRecyclerView = findViewById(R.id.message_list);
        mMsgUnread = findViewById(R.id.msg_unread);

        mEditText = findViewById(R.id.editText);
        mEmojiMenuContainer = findViewById(R.id.emojicon_menu_container);
        mMainLayout = findViewById(R.id.rootView);
        mBtnSend = findViewById(R.id.btn_send);
        mBtnSend.setOnClickListener(this);
        mBtnFace = findViewById(R.id.btnFace);
        mBtnFace.setOnClickListener(this);


        inputView = findViewById(R.id.inputView);

        btnRecord = findViewById(R.id.btnRecord);
        btnRecord.setOnClickListener(this);
        recordStateView = findViewById(R.id.recordStateView);
        recordView = findViewById(R.id.recordView);
        recordView.setSessionId(mSessionID);
        recordView.setRecordStateView(recordStateView);
        recordView.setAudioRecordListener(this);

        btnKeyboard = findViewById(R.id.btnKeyboard);
    }

    protected void init() {
        Intent intent = getIntent();
        if (intent != null) {
            mSessionID = intent.getExtras().getString("session_id");
            mBaseUser = intent.getExtras().getParcelable("user");
        }
        if (mSessionID == null || mBaseUser == null) {
            finish();
            return;
        }
    }

    private void getChatUserInfo() {
    }

    @Override
    public void onClick(View v) {
        int id  = v.getId();
        if (id == R.id.btn_send) {
            sendTxtMessage();
        }else if (id == R.id.btnFace){
            toggleEmojicon();
        }else if (id == R.id.btnRecord){
            checkPermission();
            showRecordButton(true);
            hideKeyboard();
        }else if (id == R.id.btnKeyboard){
            showRecordButton(false);
        }

    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(MsgStatusUpdateEvent event) {
        LoggerUtil.d(TAG, "MsgStatusUpdateEvent: %s", event);
        if (event.getMessage().getSessionId().equalsIgnoreCase(String.valueOf(mSessionID))) {
            mAdapter.updateData(event.getMessage());
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(ChatMsgEvent event) {
        LoggerUtil.d(TAG,"ChatMsgEvent: %s", event);
        if (event.getSessionId().equalsIgnoreCase(String.valueOf(mSessionID))) {
            if (event.getMessage().getBodyType() == BodyType.TYPE_GROUP_APPLY_NUM) {
                try {
                   // showCountHint(Integer.parseInt(event.getMessage().getContentDesc().toString()));
                } catch (Exception e) {
                }
            } else {
                if (mAdapter.getDataList().contains(event.getMessage())) {
                    mAdapter.updateData(event.getMessage());
                } else {
                    mAdapter.addData(event.getMessage());
                    if (event.getMessage().isMine() || isBottom()) {
                        scrollToBottom();
                    }
                }
            }

        }
    }

    protected boolean isBottom() {
        if (mRecyclerView == null) return false;
        if (mRecyclerView.computeVerticalScrollExtent() + mRecyclerView.computeVerticalScrollOffset()
                >= mRecyclerView.computeVerticalScrollRange())
            return true;
        return false;
    }

    private void showRecordButton(boolean isShow) {
        recordView.setVisibility(isShow ? View.VISIBLE : View.GONE);
        inputView.setVisibility(isShow ? View.GONE : View.VISIBLE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBusManager.getInstance().unRegister(this);
    }

    @Override
    public void recordFinished(String path, int time) {
        MessageManager.getInstance().sendAudioMessage(mBaseUser, path, time);
    }
}

File: mq/src/main/java/io/dourl/mqtt/ui/ChatActivity.kt
package io.dourl.mqtt.ui

import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.*
import android.widget.TextView.OnEditorActionListener
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.dourl.mqtt.R
import io.dourl.mqtt.bean.MessageModel
import io.dourl.mqtt.bean.SessionModel
import io.dourl.mqtt.bean.UserModel
import io.dourl.mqtt.constants.Constants
import io.dourl.mqtt.event.ChatMsgEvent
import io.dourl.mqtt.event.MsgStatusUpdateEvent
import io.dourl.mqtt.manager.EventBusManager
import io.dourl.mqtt.manager.MessageManager
import io.dourl.mqtt.model.message.chat.*
import io.dourl.mqtt.model.message.emoji.DefaultEmojiconDatas
import io.dourl.mqtt.model.message.emoji.Emojicon
import io.dourl.mqtt.model.message.emoji.EmojiconGroupEntity
import io.dourl.mqtt.storage.DbCallback
import io.dourl.mqtt.storage.MessageDao
import io.dourl.mqtt.storage.SessionDao
import io.dourl.mqtt.storage.SessionManager
import io.dourl.mqtt.ui.adpater.chat.*
import io.dourl.mqtt.ui.widge.*
import io.dourl.mqtt.ui.widge.EmojiconMenuBase.EmojiconMenuListener
import io.dourl.mqtt.ui.widge.RecordView.AudioRecordListener
import io.dourl.mqtt.utils.AndroidUtils
import io.dourl.mqtt.utils.AppContextUtil
import io.dourl.mqtt.utils.ImSmileUtils
import io.dourl.mqtt.utils.MediaFilesUtils
import io.dourl.mqtt.utils.log.LoggerUtil
import io.reactivex.functions.Consumer
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.io.File
import java.util.*

/**
 * File description.
 *
 * @author dourl
 * @date 2022/3/8
 */
open class ChatActivity : BaseActivity(), View.OnClickListener, AudioRecordListener {
    var mRecyclerView: RecyclerView? = null
    var noticeContent: TextView? = null
    var noticeClose: ImageView? = null
    var noticeLayout: RelativeLayout? = null
    var mTitleBar: TitleBar? = null
    var msgCountHint: TextView? = null
    var msgUnread: TextView? = null

    @JvmField
    var mEditText: MultiLineEditText? = null
    lateinit var mBtnSend: ImageView
    lateinit var btnAlbum: ImageButton
    var btnPhoto: ImageButton? = null
    var btnRecord: ImageButton? = null
    var mBtnFace: ImageButton? = null
    var mEmojiMenuContainer: FrameLayout? = null
    var inputView: LinearLayout? = null
    var btnKeyboard: ImageView? = null
    var recordView: RecordView? = null
    var bottomView: FrameLayout? = null
    var recordStateView: RecordStateView? = null
    var mMainLayout: RelativeLayout? = null

    @JvmField
    protected var mSessionID: String? = null
    private var mBaseUser: UserModel? = null
    protected var mAdapter: ChatAdapter? = null
    protected var mUnreadCount = 15
    private var mLastMsgId: Long = 0
    protected var mMsgUnread: TextView? = null
    protected var ctrlPress = false
    private var rightShown = false
    protected var mSendFile: File? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        EventBusManager.getInstance().register(this)
        init()
        initView()
        mTitleBar!!.showLine()
        mTitleBar!!.setLeftBtnListener { finish() }
        mTitleBar!!.setTitle(mSessionID)
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.orientation = LinearLayoutManager.VERTICAL
        linearLayoutManager.isAutoMeasureEnabled = true
        linearLayoutManager.stackFromEnd = true
        linearLayoutManager.reverseLayout = true
        mRecyclerView!!.layoutManager = linearLayoutManager
        mRecyclerView!!.setHasFixedSize(false)
        mAdapter = ChatAdapter(ArrayList())
        mAdapter?.register(MessageModel::class)?.to(
            ChatHintBinder(),
            ChatTextBinder(),
            ChatImageBinder(),
            ChatAudioBinder(),
            ChatVideoBinder()
        )?.withKotlinClassLinker { _, item ->
            when (item.body) {
                is HintBody -> ChatHintBinder::class
                is TextBody -> ChatTextBinder::class
                is ImageBody -> ChatImageBinder::class
                is AudioBody -> ChatAudioBinder::class
                else -> ChatVideoBinder::class
            }
        }
        mRecyclerView!!.adapter = mAdapter
        setupMessageList()
        setupInputBar()
        mRecyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (linearLayoutManager.findLastCompletelyVisibleItemPosition() == linearLayoutManager.itemCount - 1) {
                    LoggerUtil.d(TAG, "reach top")
                    //                    mPtrLayout.autoRefresh(true, 200);
                    getMsgsByPage(0)
                }
                if (mMsgUnread!!.isShown && linearLayoutManager.findLastCompletelyVisibleItemPosition() == mUnreadCount - 1) {
                    mMsgUnread!!.visibility = View.GONE
                }
            }
        })
        mRecyclerView!!.addOnLayoutChangeListener { v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom ->
            if (bottom < oldBottom - 300) {
                LoggerUtil.d(TAG, "LayoutChange oldBottom")
                mRecyclerView!!.postDelayed({ scrollToBottom() }, 100)
            }
        }
        initEmoji()
    }

    protected var emojiconMenu: EmojiconMenuBase? = null
    private fun initEmoji() {
        if (emojiconMenu == null) {
            emojiconMenu = LayoutInflater.from(this)
                .inflate(R.layout.im_layout_emojicon_menu, null) as EmojiconMenu
            val emojiconGroupList = ArrayList<EmojiconGroupEntity>()
            emojiconGroupList.add(
                EmojiconGroupEntity(
                    R.drawable.ee_1,
                    Arrays.asList(*DefaultEmojiconDatas.getData())
                )
            )
            (emojiconMenu as EmojiconMenu?)!!.init(emojiconGroupList)
        }
        mEmojiMenuContainer!!.addView(emojiconMenu)

        // emojicon menu
        emojiconMenu!!.setEmojiconMenuListener(object : EmojiconMenuListener {
            override fun onExpressionClicked(emojicon: Emojicon) {
                if (emojicon.type != Emojicon.Type.BIG_EXPRESSION) {
                    if (emojicon.emojiText != null) {
                        mEditText!!.append(
                            ImSmileUtils.getSmiledText(
                                this@ChatActivity,
                                emojicon.emojiText
                            )
                        )
                    }
                } else {
//                    sendBigExpressionMessage(emojicon.getName(), emojicon.getIdentityCode());
                }
            }

            override fun onDeleteImageClicked() {
                if (!TextUtils.isEmpty(mEditText!!.text)) {
                    val event = KeyEvent(
                        0,
                        0,
                        0,
                        KeyEvent.KEYCODE_DEL,
                        0,
                        0,
                        0,
                        0,
                        KeyEvent.KEYCODE_ENDCALL
                    )
                    dispatchKeyEvent(event)
                }
            }

            override fun onSendClicked() {
                sendTxtMessage()
            }
        })
    }

    protected fun toggleEmojicon() {
        if (mEmojiMenuContainer!!.visibility == View.VISIBLE) {
            mEmojiMenuContainer!!.visibility = View.GONE
        } else {
            sHandler.postDelayed({ mEmojiMenuContainer!!.visibility = View.VISIBLE }, 50)
            hideKeyboard()
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupInputBar() {
        mEditText!!.setOnKeyListener { v, keyCode, event -> // test on Mac virtual machine: ctrl map to KEYCODE_UNKNOWN
            if (keyCode == KeyEvent.KEYCODE_UNKNOWN) {
                if (event.action == KeyEvent.ACTION_DOWN) {
                    ctrlPress = true
                } else if (event.action == KeyEvent.ACTION_UP) {
                    ctrlPress = false
                }
            }
            false
        }
        mEditText!!.setOnEditorActionListener(OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_SEND || actionId == EditorInfo.IME_ACTION_DONE ||
                event.keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN && ctrlPress == true
            ) {
                sendTxtMessage()
                return@OnEditorActionListener true
            }
            false
        })
        mEditText!!.onFocusChangeListener = OnFocusChangeListener { view, b ->
            if (b) {
                toggleEditMode()
            }
        }
        mRecyclerView!!.setOnTouchListener { v, event ->
            if (event.action == MotionEvent.ACTION_UP) {
                hideKeyboard()
                if (rightShown) toggleRightLayout()
            }
            false
        }
    }

    protected fun toggleRightLayout() {
        var animator: ObjectAnimator? = null
        animator = if (rightShown) {
            ObjectAnimator.ofFloat(
                mMainLayout,
                "translationX",
                -AppContextUtil.dip2px(240f).toFloat(),
                0f
            )
        } else {
            ObjectAnimator.ofFloat(
                mMainLayout,
                "translationX",
                0f,
                -AppContextUtil.dip2px(240f).toFloat()
            )
        }
        animator.duration = 500
        animator.interpolator = AccelerateDecelerateInterpolator()
        animator.start()
        rightShown = !rightShown
        hideKeyboard()
    }

    protected fun hideKeyboard() {
        val focusView = currentFocus
        if (focusView != null) {
            val imm = focusView.context.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(focusView.windowToken, 0)
            focusView.clearFocus()
        }
    }

    private fun toggleEditMode() {
        sHandler.postDelayed({ mEmojiMenuContainer!!.visibility = View.GONE }, 50)
    }

    protected open fun sendTxtMessage() {
        val text = mEditText!!.text.toString()
        if (text.length > 0) {
            MessageManager.getInstance().sendTextMessage(mBaseUser, text)
            mEditText!!.setText("")
            scrollToBottom()
        } else {
            // sendEmptyTxtMessage();
        }
        if (!mBaseUser!!.isFriend) {
            /*HINT_LIMIT--;
            if (HINT_LIMIT == -1) {
                new ChatHintDialog(this, mMainLayout, new ChatHintDialog.OnFunClickListener() {
                    @Override
                    public void onFunClick() {
                        showLoadingProgress();
                        doContactsAction();
                    }
                }).show();
            }*/
        }
    }
    protected open fun sendImageMessage(path: String?) {
        MessageManager.getInstance().sendImageMessage(mBaseUser, path)
    }

    protected fun scrollToBottom() {
        mRecyclerView!!.scrollToPosition(0)
    }

    private fun setupMessageList() {
        SessionDao.findSessionById(mSessionID, object : DbCallback<SessionModel?> {
            override fun onSuccess(model: SessionModel?) {
                runOnUiThread {
                    if (model != null) {
                        mUnreadCount = model.unreadMsgCount
                        if (mUnreadCount > 15) {
                            mMsgUnread!!.visibility = View.VISIBLE
                            mMsgUnread!!.text =
                                getString(R.string.hint_chat_unread_msg, mUnreadCount)
                        }
                        mEditText!!.setText(model.draft)
                    }
                    getMsgsByPage(mUnreadCount)
                    SessionManager.getInstance().resetUnreadCount(mSessionID)
                }
            }

            override fun onFail(e: Throwable) {}

        })
    }

    @SuppressLint("CheckResult")
    private fun getMsgsByPage(unread: Int) {
        MessageDao.getMsgsBySessionIdObservable(mSessionID, mLastMsgId, PRE_PAGE_COUNT + unread)
            .subscribe(
                Consumer { messagePaging ->
                    if (messagePaging == null) return@Consumer
                    if (0L == mLastMsgId) {
                        mAdapter!!.setData(messagePaging.list)
                        scrollToBottom()
                        //                    mRecyclerView.scrollToPosition(messagePaging.list.size() - 1);
                    } else {
                        mAdapter!!.addPreviousPage(messagePaging.list)
                    }
                    mLastMsgId = messagePaging.lastMsgDbId
                    //mPtrLayout.setPullToRefresh(messagePaging.hasMore);
                })
    }

    protected fun initView() {
        mTitleBar = findViewById(R.id.titleBar)
        mRecyclerView = findViewById(R.id.message_list)
        mMsgUnread = findViewById(R.id.msg_unread)
        mEditText = findViewById(R.id.editText)
        mEmojiMenuContainer = findViewById(R.id.emojicon_menu_container)
        mMainLayout = findViewById(R.id.rootView)
        mBtnSend = findViewById(R.id.btn_send)
        mBtnSend?.setOnClickListener(this)
        mBtnFace = findViewById(R.id.btnFace)
        mBtnFace?.setOnClickListener(this)
        inputView = findViewById(R.id.inputView)
        btnRecord = findViewById(R.id.btnRecord)
        btnRecord?.setOnClickListener(this)
        recordStateView = findViewById(R.id.recordStateView)
        recordView = findViewById(R.id.recordView)
        recordView?.setSessionId(mSessionID)
        recordView?.setRecordStateView(recordStateView)
        recordView?.setAudioRecordListener(this)
        btnKeyboard = findViewById(R.id.btnKeyboard)
        btnAlbum = findViewById(R.id.btnAlbum)
        btnAlbum.setOnClickListener(this)
    }

    protected open fun init() {
        val intent = intent
        if (intent != null) {
            mSessionID = intent.extras!!.getString("session_id")
            mBaseUser = intent.extras!!.getParcelable("user")
        }
        if (mSessionID == null || mBaseUser == null) {
            finish()
            return
        }
    }

    private val chatUserInfo: Unit
        private get() {}

    override fun onClick(v: View) {
        val id = v.id
        if (id == R.id.btn_send) {
            sendTxtMessage()
        } else if (id == R.id.btnFace) {
            toggleEmojicon()
        } else if (id == R.id.btnRecord) {
            checkPermission()
            showRecordButton(true)
            hideKeyboard()
        } else if (id == R.id.btnKeyboard) {
            showRecordButton(false)
        } else if (id == R.id.btnAlbum){
            choosePhoto();
        }
    }

    private fun choosePhoto() {
        mSendFile = MediaFilesUtils.getImageFile(this);
        AndroidUtils.openAlbum(this, mSendFile, false, 0, 0, Constants.REQUEST_CODE_OPEN_ALBUM, Constants.FILE_PROVIDER);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: MsgStatusUpdateEvent) {
        LoggerUtil.d(TAG, "MsgStatusUpdateEvent: %s", event)
        if (event.message.sessionId.equals(mSessionID.toString(), ignoreCase = true)) {
            mAdapter!!.updateData(event.message)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEventMainThread(event: ChatMsgEvent) {
        LoggerUtil.d(TAG, "ChatMsgEvent: %s", event)
        if (event.sessionId.equals(mSessionID.toString(), ignoreCase = true)) {
            if (event.message.bodyType == BodyType.TYPE_GROUP_APPLY_NUM) {
                try {
                    // showCountHint(Integer.parseInt(event.getMessage().getContentDesc().toString()));
                } catch (e: Exception) {
                }
            } else {
                if (mAdapter!!.dataList.contains(event.message)) {
                    mAdapter!!.updateData(event.message)
                } else {
                    mAdapter!!.addData(event.message)
                    if (event.message.isMine() || isBottom) {
                        scrollToBottom()
                    }
                }
            }
        }
    }

    protected val isBottom: Boolean
        protected get() {
            if (mRecyclerView == null) return false
            return if (mRecyclerView!!.computeVerticalScrollExtent() + mRecyclerView!!.computeVerticalScrollOffset()
                >= mRecyclerView!!.computeVerticalScrollRange()
            ) true else false
        }

    private fun showRecordButton(isShow: Boolean) {
        recordView!!.visibility = if (isShow) View.VISIBLE else View.GONE
        inputView!!.visibility = if (isShow) View.GONE else View.VISIBLE
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBusManager.getInstance().unRegister(this)
    }

    override fun recordFinished(path: String, time: Int) {
        MessageManager.getInstance().sendAudioMessage(mBaseUser, path, time)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_OK) {
            return;
        }
        if (requestCode == Constants.REQUEST_CODE_OPEN_ALBUM) {
            if (mSendFile != null) {
             val filePath =   AndroidUtils.parseActivityMediaResult(this, mSendFile, requestCode, resultCode, data, Constants.REQUEST_CODE_TAKE_PHOTO, Constants.REQUEST_CODE_OPEN_ALBUM);
                if (filePath != null) {
                    sendImageMessage(filePath);
                }
            }
        }
    }



    companion object {
        private const val TAG = "ChatActivity"
        fun intentTo(context: Context, sessionId: String?) {
            val intent = Intent(context, ChatActivity::class.java)
            val bundle = Bundle()
            bundle.putString("session_id", sessionId)
            intent.putExtras(bundle)
            context.startActivity(intent)
        }

        @JvmStatic
        fun intentTo(context: Context, sessionId: String?, user: UserModel?) {
            val intent = Intent(context, ChatActivity::class.java)
            val bundle = Bundle()
            bundle.putString("session_id", sessionId)
            bundle.putParcelable("user", user)
            intent.putExtras(bundle)
            context.startActivity(intent)
        }

        const val PRE_PAGE_COUNT = 10 //最少10
    }
}